onnxruntime/core/providers/coreml/builders/impl/resize_op_builder.cc

```diff
@@ -0,0 +1,273 @@
+  // since there is no use of it, we will not process it later
+  model_builder.AddInitializerToSkip(node.InputDefs()[2]->Name());  // scales
+  model_builder.AddInputToSkip(node.InputDefs()[2]->Name());        // scales
+
+  if (node.InputDefs().size() > 3) {
+    model_builder.AddInitializerToSkip(node.InputDefs()[3]->Name());  // sizes
+    model_builder.AddInputToSkip(node.InputDefs()[3]->Name());        // sizes
+  }
+}
+
+Status ResizeOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder,
+                                              const Node& node,
+                                              const logging::Logger& logger) const {
+  std::unique_ptr<COREML_SPEC::NeuralNetworkLayer> layer = CreateNNLayer(node);
+
+  auto* coreml_upsample = layer->mutable_upsample();
+  NodeAttrHelper helper(node);
+  const auto mode = helper.Get("mode", "nearest");
+  if (mode == "linear") {
+    coreml_upsample->set_mode(COREML_SPEC::UpsampleLayerParams_InterpolationMode_BILINEAR);
+  } else {  // we already checked the mode must be NN or Bilinear in IsOpSupportedImpl
+    coreml_upsample->set_mode(COREML_SPEC::UpsampleLayerParams_InterpolationMode_NN);
+  }
+
+  const auto& input_defs = node.InputDefs();
+  const auto& initializers(model_builder.GetInitializerTensors());
+
+  if (input_defs.size() == 3) {  // use scales
+    std::vector<float> scales;
+    ORT_RETURN_IF_NOT(GetResizeScales(initializers, node, scales), "Error getting resize scales");
+    coreml_upsample->add_scalingfactor(static_cast<int64_t>(scales[2]));
+    coreml_upsample->add_scalingfactor(static_cast<int64_t>(scales[3]));
+  } else {  // we already checked number of inputs in IsOpSupportedImpl
+    std::vector<int64_t> input_shape;
+    ORT_RETURN_IF_NOT(GetShape(*input_defs[0], input_shape, logger), "Error getting input shape");
+    std::vector<int64_t> output_sizes;
+    ORT_RETURN_IF_NOT(GetResizeOutputSizes(initializers, node, output_sizes), "Error getting resize output_sizes");
+    coreml_upsample->add_scalingfactor(static_cast<int64_t>(output_sizes[2] / input_shape[2]));
+    coreml_upsample->add_scalingfactor(static_cast<int64_t>(output_sizes[3] / input_shape[3]));
+  }
+
+  *layer->mutable_input()->Add() = input_defs[0]->Name();
+  *layer->mutable_output()->Add() = node.OutputDefs()[0]->Name();
+
+  model_builder.AddLayer(std::move(layer));
+  return Status::OK();
+}
+
+// Operator support related
+
+bool ResizeOpBuilder::IsOpSupportedImpl(const InitializedTensorSet& initializers, const Node& node,
+                                        const logging::Logger& logger) const {
+  const auto& input_defs = node.InputDefs();
+
+  std::vector<int64_t> input_shape;
+  if (!GetShape(*input_defs[0], input_shape, logger))
+    return false;
+
+  const auto input_size = input_shape.size();
+  if (input_size != 4) {
+    LOGS(logger, VERBOSE) << "Resize only support 4d shape, input is "
+                          << input_size << "d shape";
+    return false;
+  }
+
```

Comment:
When we create the mobile friendly model do we ever remove the batch dim so a 3d shape is possible? Or it just gets locked to being 1?

Reply:
By convention, 2d image has 4d shape, 3d shape is of something like NCH which is 1d array/image